src/relentless/optimize/objective.py

```diff
@@ -375,6 +384,65 @@ def compute(self, variables, directory=None):
         return result
 
     def compute_gradient(self, ensemble, variables):
+        """Computes the relative entropy gradient for an ensemble.
+
+        Parameters
+        ----------
+        ensemble : :class:`~relentless.ensemble.Ensemble`
+            The ensemble for which to evaluate the gradient.
+        variables : :class:`~relentless.variable.Variable` or tuple
+            Variables with respect to which to compute gradient.
+
+        Returns
+        -------
+        :class:`~relentless.math.KeyedArray`
+            The gradient, keyed on the ``variables``.
+
+        """
+        if isinstance(self.target, Ensemble) and isinstance(
+            self.thermo, EnsembleAverage
+        ):
+            return self._compute_gradient_old(ensemble, variables)
+        else:
+            return self._compute_gradient_new(ensemble, variables)
+
+    def _compute_gradient_new(self, ensemble, variables):
+        r"""Computes the relative entropy gradient for an ensemble.
+
+        Parameters
+        ----------
+        ensemble : :class:`string`
+            Path to gsd trajectory file.
+        variables : :class:`~relentless.variable.Variable` or tuple
+            Variables with respect to which to compute gradient.
+
+        Returns
+        -------
+        :class:`~relentless.math.KeyedArray`
+            The gradient, keyed on the ``variables``.
+
+        """
+
+        # load the target ensemble trajectory file
+        traj_tgt = self.target
+        traj_sim = ensemble
+        dvars = variable.graph.check_variables_and_types(variables, variable.Variable)
+        gradient = math.KeyedArray(keys=dvars)
+
+        # putting these lines in so that I can push this commit, please ignore
+        traj_sim, traj_tgt = 0, 0
+        update = traj_sim + traj_tgt
+
+        for var in dvars:
+            update = 0
+            types = collections.PairMatrix(
+                keys=self.potentials.pair.potentials[0].coeff.types
+            )
+            for i, j in types:
+                gradient[var] = update
```

Comment:
Here I think we should pull the types from the loaded files and make a pair matrix and keyed array for all the bonded types then loop through evaluating. Probably also need a check to make sure the the types are the same between the two files. 

Let me know if you think there is a better way to implement this.

Reply:
Yes, I think that it would make sense to extract the types from the trajectory files for all interactions (pairs, bonds, etc.). Line 439 above will have an issue if you don't do this because it's possible for different potentials to be keyed on different types.